src/content/docs/en/guides/actions.mdx

```diff
@@ -560,24 +550,155 @@ const result = Astro.getActionResult(actions.addToCart);
 <!--...-->
 ```
 
-:::caution
-Action data is passed using a persisted cookie. **This cookie is not encrypted** and is limited to 4 KB in size, although the exact limit may vary between browsers. 
+### Advanced: Persist action results with a session
 
-In general, we recommend returning the minimum information required from your action `handler` to avoid vulnerabilities, and persist other sensitive information in a database.
+<p><Since v="5.0.0" /></p>
 
-For example, you might return the name of a product in an `addToCart` action, rather than returning the entire `product` object:
+Action results are displayed as a POST submission. This means that the result will be reset to `undefined` when a user closes and revisits the page. The user will also see a "confirm form resubmission?" dialog if they attempt to refresh the page.
 
-```ts title="src/actions/index.ts" del={7} ins={8}
-import { defineAction } from 'astro:actions';
+To customize this behavior, you can add middleware to handle the result of the action manually. You may choose the persist the action result using a cookie or session storage.
+
+Start by [creating a middleware file](https://docs.astro.build/en/guides/middleware/) and importing [the `getActionContext()` utility](/en/reference/modules/astro-actions/#getactioncontext) from `astro:actions`. This function returns an `action` object with information about the incoming action request, including the action handler and whether the action was called from an HTML form. `getActionContext()` also returns the `setActionResult()` and `serializeActionResult()` functions to programmatically set the value returned by `Astro.getActionResult()`:
+
+```ts title="src/middleware.ts" {2,5}
+import { defineMiddleware } from 'astro:middleware';
+import { getActionContext } from 'astro:actions';
+
+export const onRequest = defineMiddleware(async (context, next) => {
+  const { action, setActionResult, serializeActionResult } = getActionContext(context);
+  if (action?.calledFrom === 'form') {
+    const result = await action.handler();
+    // ... handle the action result
+    setActionResult(action.name, serializeActionResult(result));
+  }
+  return next();
+});
+```
+
+A common practice to persist HTML form results is the [POST / Redirect / GET pattern](https://en.wikipedia.org/wiki/Post/Redirect/Get). This approach removes "confirm form resubmission?" dialog when a user refreshes by using a redirect, and gives you the option to persist action results throughout the user's session.
+
+This example applies the POST / Redirect / GET pattern to all form submissions using session storage. Action results are written to a session store using [Netlify Blob](https://docs.netlify.com/blobs/overview/), and retrieved after a redirect using a session ID:
+
+:::note
+This example requires the [Netlify server adapter](https://docs.astro.build/en/guides/integrations-guide/netlify/) to run. Feel free to adapt this example to your storage platform of choice.
```

Comment:
Don't be fooled by this being outdated just because the link was updated! I suggest removing the note and working the context in normal paragraph text!